base/sysinfo.jl

```diff
@@ -321,4 +323,91 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys.isexecutable(path::AbstractString)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function isexecutable(path::AbstractString)
+    @static if iswindows()
+        return isfile(path)
+    else
+        # We use `access()` and `X_OK` to determine if a given path is
+        # executable by the current user.  `X_OK` comes from `unistd.h`.
+        X_OK = 0x01
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+
+"""
+    Sys.which(program_name::AbstractString)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns the absolute
+path. Raises `ArgumentError` if no such program is available.  If a relative or
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::AbstractString)
```

Comment:
Probably 
```jl
which(program_name::AbstractString) = which(String(program_name))
function which(program_name::String)
    ...
end
```
as otherwise `program_name` will get converted to `String` multiple times if it is some other string type.